recipe_scrapers/_grouping_utils.py

```diff
@@ -0,0 +1,137 @@
+    first : str
+        First string
+    second : str
+        Second string
+
+    Returns
+    -------
+    float
+        Description
+    """
+
+    if first == second:
+        # Indentical sentences have maximum score of 1
+        return 1
+
+    if len(first) < 2 or len(second) < 2:
+        # If either sentence has 0 or 1 character we can't generate bigrams,
+        # so the score is 0
+        return 0
+
+    firest_bigrams = set([first[i : i + 2] for i in range(len(first) - 1)])
+    second_bigrams = set([second[i : i + 2] for i in range(len(second) - 1)])
+
+    intersection = firest_bigrams & second_bigrams
+
+    return 2.0 * len(intersection) / (len(firest_bigrams) + len(second_bigrams))
+
+
+def best_match(test_string: str, target_strings: List[str]) -> str:
+    """Find the best match for test_string in target_strings
+
+    Parameters
+    ----------
+    test_string : str
+        Test string to find the best match for.
+    target_strings : List[str]
+        List of target strings to search for best match in
+
+    Returns
+    -------
+    str
+        The closest matching string in target_strings to test_string
+
+    """
+    scores = [
+        score_sentence_similarity(test_string, target) for target in target_strings
+    ]
+    # Get the index of the maximum score
+    index, _ = max(enumerate(scores), key=lambda x: x[1])
+
+    return target_strings[index]
+
+
+def group_ingredients(
+    ingredients_list: List[str],
+    soup: BeautifulSoup,
+    group_heading: str,
+    group_element: str,
+) -> List[IngredientGroup]:
+    """
+    Group ingredients into sublists according the heading in the recipe.
+
+    The group headings are determined by html element matching the group_heading CSSselector.
+    The group items are determined by extracting the text the html elements that are
+    below the group heading, then matching the text to one of the ingredients in ingredients_list.
+    This is done to ensure this function returns the same ingredients as are returned from
+    the .ingredients() method of a scraper.
+
+    If the recipe doesn't group ingredients, this returns a single IngredientGroup object
+    with the purpose set to None and the ingredients list populated with all ingredients.
+
+    Parameters
+    ----------
+    ingredients_list : List[str]
+        List of ingredients extracted by recipe scraper
+    soup : TYPE
+        Description
+    group_heading : str
+        CSS selector for ingredient group heading
+    group_element : str
+        CSS selector for ingredient elements
+
+    Returns
+    -------
+    List[IngredientGroup]
+        List of IngredientGroup objects.
+        Each object represents a group of ingredients and their purpose.
+
+    """
+    groupings: Dict[Optional[str], List[str]] = {None: []}
+    current_heading = None
+
+    # Find all elemement matching the heading and ingredient selectors
+    elements = soup.select(",".join([group_heading, group_element]))
```

Comment:
What's the reason to iterate over both grouping items and ingredient items within the same loop construct?  (I think that's what this does?  please correct me if I'm mistaken)

Reply:
group_heading is the selector for just the text of the heading, group_element is the selector for the text of each ingredient. By selecting both together, we get a list of all the group headings and ingredients in the order in which they appear in the html e.g. `[heading, ing, ing, ing, heading, ing, ing, heading, ....]`
The loop after line 119 is there to turn the list into a structure by assigning each ingredient to the heading that appears before it in the list.

Reply:
Ok, that seems reasonable.

I'm vaguely wondering about situations where the HTML of the page might include unpredictable orderings of group heading elements compared to ingredient content elements.  In situations like that, I think it could make sense to use a two-pass approach: first pass to collect the headings, second pass to collect the ingredients, and then some procedural code to combine the two sets.